b2sdk/utils/wildcards.py

```diff
@@ -0,0 +1,116 @@
+    max_index = len(folder_to_list)
+    while offset < max_index:
+        starter_index = folder_to_list.index(wildcard_character, offset)
+        if starter_index is None:
+            return None
+        elif starter_index > 0 and folder_to_list[starter_index - 1] == '\\':
+            # the character is escaped, ignore it
+            offset = starter_index + 1
+            continue
+        return starter_index
+    raise ValueError("no unescaped character found")
+
+
+def get_solid_prefix(
+    current_prefix: str, folder_to_list: str, wildcard_style: WildcardStyle
+) -> str:
+    """If we're running with wildcard-matching, we could get a different prefix from it.
+    We search for the first occurrence of the special characters and fetch parent path from that place.
+
+    Examples:
+       'b/c/*.txt' –> 'b/c/'
+       '*.txt' –> ''
+       'a' –> 'a/'
+    """
+    MATCHERS = {
+        # wildcard style: (wildcard match checker, allowed wildcard chars)
+        WildcardStyle.SHELL.value:
+            (
+                _find_unescaped_char,
+                ('*', '?', '[', '{'),  # ** is matched via *
+            ),
+        WildcardStyle.GLOB.value: (
+            lambda folder, char: folder.index(char),
+            ('*', '?', '['),
+        ),
+    }
+
+    try:
+        finder, charset = MATCHERS[wildcard_style]
+    except KeyError:
+        raise ValueError(f'Unknown wildcard style: {wildcard_style!r}')
+
+    solid_length = len(folder_to_list)
+    for wildcard_character in charset:
+        try:
+            char_index = finder(folder_to_list, wildcard_character)
+        except ValueError:
+            continue
+        else:
+            solid_length = min(char_index, solid_length)
+
+    # +1 to include the starter character.  Using posix path to
+    # ensure consistent behaviour on Windows (e.g. case sensitivity).
+    path = pathlib.PurePosixPath(folder_to_list[:solid_length + 1])
+    parent_path = str(path.parent)
+
+    # Path considers dot to be the empty path.
+    # There's no shorter path than that.
+    if parent_path == '.':
+        return ''
+
+    # We could receive paths in different stage, e.g. 'a/*/result.[ct]sv' has two
+    # possible parent paths: 'a/' and 'a/*/', with the first one being the correct one
+    return min(parent_path, current_prefix, key=len)
+
+
+def get_wildcard_matcher(match_pattern: str,
+                         wildcard_style: WildcardStyle) -> Callable[[str], bool]:
+    """Return a wildcard matcher for chosen style and pattern."""
+    if wildcard_style == WildcardStyle.SHELL:
+        wc_flags = (
+            wcglob.CASE  # case sensitive
+            | wcglob.BRACE  # support {} for multiple options
+            | wcglob.GLOBSTAR  # support ** for recursive matching
+            | wcglob.NEGATE  # support [!] for negation
+        )
+        wildcard_matcher = partial(
+            lambda file_name: wcglob.globmatch(file_name, match_pattern, flags=wc_flags, limit=100)
```

Comment:
This limit (lowered from 1000 to 100) is hardcoded and undocumented. I'm afraid it will bite someone in the ass one day. Can we set it to a value that won't ever trigger (with the limitation of file name / object length) or expose it so that the user can increase it if they have a weird use case? Otherwise sooner or later someone will hit it, report a bug and we'll have to figure out a way to fix it then. Why did you decide to set it to a 100?

Reply:
> This limit (lowered from 1000 to 100) is hardcoded and undocumented

using `wcmatch` can be considered an implementation detail, in which case even the limit 1000 is undocumented/hardcoded.

> Why did you decide to set it to a 100?

the docs had a warning there "Massive Expansion Risk" which made me think what would be a common use-case this feature might be covering - either expanding a-z, or expanding 0-99. I went with the latter one. 
I dont have strong opinion about this, so I'll revert it to 1000, which sounds loose enough.